rust/src/sdp/parser.rs

```diff
@@ -0,0 +1,694 @@
+        tuple((
+            map_res(take_while(is_time_char), std::str::from_utf8),
+            space1,
+            map_res(take_while(is_time_char), std::str::from_utf8),
+            space1,
+            map_res(take_while(is_time_char), std::str::from_utf8),
+            space1,
+            map_res(take_while(is_time_char), std::str::from_utf8),
+        )),
+    )(i)?;
+    let (i, _) = line_ending(i)?;
+    let tz = format!("{} {} {} {}", z1, z2, z3, z4);
+    Ok((i, tz.to_string()))
+}
+
+fn parse_encryption_key(i: &[u8]) -> IResult<&[u8], String> {
+    let (i, key) = preceded(
+        tag("k="),
+        map_res(take_till(is_line_ending), std::str::from_utf8),
+    )(i)?;
+    let (i, _) = line_ending(i)?;
+    Ok((i, key.to_string()))
+}
+
+fn parse_attributes(i: &[u8]) -> IResult<&[u8], Vec<String>> {
+    let (i, attrs) = many0(preceded(
+        tag("a="),
+        tuple((
+            map_res(take_while(is_alphabetic), std::str::from_utf8),
+            opt(preceded(
+                char_parser(':'),
+                map_res(take_till(is_line_ending), std::str::from_utf8),
+            )),
+            line_ending,
+        )),
+    ))(i)?;
+    let vec = attrs
+        .iter()
+        .map(|a| {
+            if let Some(val) = a.1 {
+                format!("{}:{}", a.0, val)
+            } else {
+                a.0.to_string()
+            }
+        })
+        .collect();
+    Ok((i, vec))
+}
+
+fn parse_media_description(i: &[u8]) -> IResult<&[u8], MediaDescription> {
+    let (i, _) = tag("m=")(i)?;
+    let (i, media) = map_res(
+        alt((
+            tag("audio"),
+            tag("video"),
+            tag("text"),
+            tag("application"),
+            tag("message"),
+        )),
+        |bytes: &[u8]| String::from_utf8(bytes.to_vec()),
+    )(i)?;
+    let (i, _) = space1(i)?;
+
+    let (i, port) = map_res(
+        take_while_m_n(1, 5, |b: u8| b.is_ascii_digit()),
+        std::str::from_utf8,
+    )(i)?;
+    let (i, number_of_ports) = opt(preceded(
+        char_parser('/'),
+        map_res(
+            take_while_m_n(1, 5, |b: u8| b.is_ascii_digit()),
+            std::str::from_utf8,
+        ),
+    ))(i)?;
+    let (i, _) = space1(i)?;
+
+    let (i, proto) = map_res(
+        alt((tag("udp"), tag("RTP/AVP"), tag("RTP/SAVP"))),
+        |bytes: &[u8]| String::from_utf8(bytes.to_vec()),
+    )(i)?;
+
+    // Parse fmt list
+    let (i, fmt) = many1(preceded(
+        space1,
+        map_res(
+            take_while_m_n(1, 255, |b: u8| b.is_ascii_alphanumeric()),
+            std::str::from_utf8,
+        ),
+    ))(i)?;
+    let (i, _) = line_ending(i)?;
+
+    let (i, session_info) = opt(parse_session_info)(i)?;
+    let (i, connection_data) = opt(parse_connection_data)(i)?;
+    let (i, bandwidths) = opt(parse_bandwidth)(i)?;
+    let (i, encryption_key) = opt(parse_encryption_key)(i)?;
+    let (i, attributes) = opt(parse_attributes)(i)?;
+
+    let port = port.parse::<u16>().unwrap();
```

Comment:
Would not this panic if I supply `70000` ?

Reply:
Yes, indeed.

Reply:
This is bad :-p

Reply:
Already fixed :)